### PR TITLE
Us003 add user sessions support

### PR DESCRIPTION
**Pull request by Kost927**

### CODE REVIEW

File: src/app/shared/services/auth.service.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    return localStorage.getItem('token');
+  }
+
+  registration(newUser: NewUser) {
+    if (localStorage.allUsers) {
+      let allUsers = [...JSON.parse(localStorage.getItem('allUsers'))];
+      allUsers = [...allUsers, newUser];
+      localStorage.setItem('allUsers', JSON.stringify(allUsers));
```
Comment (JnClark88) [inline src/app/shared/services/auth.service.ts:27]: можно обойтись без лета, стараемся всегда константы использовать

File: src/app/shared/services/auth.service.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    const fetchLogin$ = new Observable((observer) => {
+      setTimeout(() => {
+        observer.next(1);
+      }, 1000);
+    });
+
+    return fetchLogin$.pipe(tap(this.setToken));
+  }
+
+  logout() {
```
Comment (JnClark88) [inline src/app/shared/services/auth.service.ts:43]: 🌴

File: src/app/successfully-registered/successfully-registered.component.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+@Component({
+  selector: 'app-successfully-registered',
+  templateUrl: './successfully-registered.component.html',
+  styleUrls: ['./successfully-registered.component.scss'],
+})
+export class SuccessfullyRegisteredComponent implements OnInit {
+  constructor(private router: Router) {}
+
+  ngOnInit(): void {}
```
Comment (JnClark88) [inline src/app/successfully-registered/successfully-registered.component.ts:12]: 🌴

File: src/app/successfully-registered/successfully-registered.component.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+  goToLoginPage() {
```
Comment (JnClark88) [inline src/app/successfully-registered/successfully-registered.component.ts:14]: 🌴

File: src/app/test-component/test-component.component.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+@Component({
+  selector: 'app-test-component',
+  templateUrl: './test-component.component.html',
+  styleUrls: ['./test-component.component.scss'],
+})
+export class TestComponentComponent implements OnInit {
+  constructor(private auth: AuthService, private router: Router) {}
+
+  ngOnInit(): void {}
```
Comment (JnClark88) [inline src/app/test-component/test-component.component.ts:13]: 🌴

File: src/app/test-component/test-component.component.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+  logout() {
```
Comment (JnClark88) [inline src/app/test-component/test-component.component.ts:15]: 🌴

File: src/app/expire-msg/expire-msg.component.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+export class ExpireMsgComponent implements OnInit {
+  expireMsg: string;
+
+  constructor(private route: ActivatedRoute, private router: Router) {}
+
+  ngOnInit(): void {
+    this.route.queryParams.subscribe((params: Params) => {
+      if (params['loginExpire']) {
+        this.expireMsg =
+          'Your current session has expired. Please login again to continue using this app!';
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/expire-msg/expire-msg.component.ts:19]: done

File: src/app/expire-msg/expire-msg.component.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      }
+    });
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/expire-msg/expire-msg.component.ts:21]: done

File: src/app/expire-msg/expire-msg.component.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  goToLoginPage() {
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/expire-msg/expire-msg.component.ts:24]: done

File: src/app/login-page/login-page.component.html [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+          type="text"
+          placeholder="Please enter your email"
+          class="input-field"
+          autocomplete="on"
+          formControlName="email"
+        />
+      </div>
+      <div
+        class="validation"
+        *ngIf="loginForm.get('email').invalid && loginForm.get('email').touched"
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.html:18]: done

File: src/app/login-page/login-page.component.html [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      >
+        <small
+          class="validate-msg"
+          *ngIf="loginForm.get('email').errors.required"
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.html:22]: done

File: src/app/login-page/login-page.component.html [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+          class="input-field"
+          autocomplete="on"
+          formControlName="password"
+        />
+      </div>
+      <div
+        class="validation"
+        *ngIf="
+          loginForm.get('password').invalid && loginForm.get('password').touched
+        "
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.html:51]: done

File: src/app/login-page/login-page.component.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+      email: new FormControl('', [
+        Validators.required,
+        Validators.email,
+        LoginValidate.existEmail,
+      ]),
+      password: new FormControl('', [
+        Validators.required,
+        LoginValidate.existPassword,
+      ]),
+    });
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.ts:31]: done

File: src/app/login-page/login-page.component.ts [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  selector: 'app-login-page',
+  templateUrl: './login-page.component.html',
+  styleUrls: ['./login-page.component.scss'],
+})
+export class LoginPageComponent implements OnInit {
+  loginForm: FormGroup;
+  signIn = false;
+
+  constructor(private router: Router, private auth: AuthService) {}
+  ngOnInit(): void {
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.ts:20]: done

File: src/app/login-page/login-page.component.ts [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        LoginValidate.existEmail,
+      ]),
+      password: new FormControl('', [
+        Validators.required,
+        LoginValidate.existPassword,
+      ]),
+    });
+  }
+
+  navigateToRegisterPage() {
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.ts:34]: done

File: src/app/login-page/login-page.component.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    this.router.navigate(['/register']);
+  }
+
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.ts:37]: done

File: src/app/login-page/login-page.component.ts [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    if (this.loginForm.invalid) {
+      return;
+    }
+
+    this.signIn = true;
+
+    const user: User = {
+      email: this.loginForm.value.email,
+      password: this.loginForm.value.password,
+    };
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.ts:48]: done

File: src/app/login-page/login-page.component.ts [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+    const allUsers = JSON.parse(localStorage.getItem('allUsers'));
+    allUsers.forEach((el) => {
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.ts:51]: done

File: src/app/login-page/login-page.component.ts [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+      if (
+        el.email === this.loginForm.value.email &&
+        el.password === this.loginForm.value.password
+      ) {
+        this.auth.login(user).subscribe((value) => {
+          this.loginForm.reset();
+          this.router.navigate(['/test']);
+        });
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.ts:59]: done

File: src/app/login-page/login-page.component.ts [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    const user: User = {
+      email: this.loginForm.value.email,
+      password: this.loginForm.value.password,
+    };
+
+    const allUsers = JSON.parse(localStorage.getItem('allUsers'));
+    allUsers.forEach((el) => {
+      if (
+        el.email === this.loginForm.value.email &&
+        el.password === this.loginForm.value.password
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/login-page/login-page.component.ts:54]: done

File: src/app/register-page/register-page.component.html [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+          id="name"
+          type="text"
+          placeholder="Please enter your name"
+          formControlName="name"
+          autocomplete="on"
+        />
+      </div>
+      <div
+        class="validation"
+        *ngIf="
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:18]: done

File: src/app/register-page/register-page.component.html [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+          class="validate-msg"
+          *ngIf="registerForm.get('name').errors.pattern"
+          >No special symbols or numbers allowed.</small
+        >
+        <small
+          class="validate-msg"
+          *ngIf="registerForm.get('name').errors.minlength"
+        >
+          Should be at least
+          {{ registerForm.get("name").errors.minlength.requiredLength }}
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:32]: done

File: src/app/register-page/register-page.component.html [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+          characters length. Now
+          {{ registerForm.get("name").errors.minlength.actualLength }}
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:34]: done

File: src/app/register-page/register-page.component.html [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+          placeholder="Please enter your email"
+          formControlName="email"
+          autocomplete="on"
+        />
+      </div>
+      <div
+        class="validation"
+        *ngIf="
+          registerForm.get('email').invalid && registerForm.get('email').touched
+        "
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:61]: done

File: src/app/register-page/register-page.component.html [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+          placeholder="Please enter your password"
+          formControlName="password"
+          autocomplete="on"
+        />
+      </div>
+      <div
+        class="validation"
+        *ngIf="
+          registerForm.get('password').invalid &&
+          registerForm.get('password').touched
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:99]: done

File: src/app/register-page/register-page.component.html [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+          class="validate-msg"
+          *ngIf="registerForm.get('password').errors.required"
+          >This field is required.</small
+        >
+        <small
+          class="validate-msg"
+          *ngIf="registerForm.get('password').errors.minlength"
+        >
+          Should be at least
+          {{ registerForm.get("password").errors.minlength.requiredLength }}
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:112]: done

File: src/app/register-page/register-page.component.html [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+          characters length. Now
+          {{ registerForm.get("password").errors.minlength.actualLength }}
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:114]: done

File: src/app/register-page/register-page.component.html [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+          characters.</small
+        >
+        <small
+          class="validate-msg"
+          *ngIf="registerForm.get('password').errors.pattern"
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:119]: done

File: src/app/register-page/register-page.component.html [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+
+        <small
+          class="validate-msg"
+          *ngIf="
+            registerForm.get('password').errors.emailIncludes ||
+            registerForm.get('password').errors.firstNameWithSpaceIncludes ||
+            registerForm.get('password').errors.secondNameWithSpaceIncludes ||
+            registerForm.get('password').errors.firstNameWithDashIncludes ||
+            registerForm.get('password').errors.secondNameWithDashIncludes
+          "
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.html:133]: done

File: src/app/register-page/register-page.component.ts [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+        Validators.email,
+        Validators.pattern(constants.EMAIL_VALIDATE),
+      ]),
+      password: new FormControl('', [
+        Validators.required,
+        Validators.minLength(5),
+        Validators.pattern(constants.PASSWORD_VALIDATE),
+        PasswordValidate.uniqPassword,
+      ]),
+    });
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.ts:37]: done

File: src/app/register-page/register-page.component.ts [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+  submit(): void {
+    if (this.registerForm.invalid) {
+      return;
+    }
+
+    const newUser: NewUser = {
+      id: Date.now(),
+      email: this.registerForm.value.email,
+      password: this.registerForm.value.password,
+      name: this.registerForm.value.name,
```
[the remark this answers is not in the document]
Comment (Kost927) [reply src/app/register-page/register-page.component.ts:49]: done

File: src/app/successfully-registered/successfully-registered.component.ts (hunk 3 above)
Comment (Kost927) [reply src/app/successfully-registered/successfully-registered.component.ts:12]: done

File: src/app/successfully-registered/successfully-registered.component.ts (hunk 4 above)
Comment (Kost927) [reply src/app/successfully-registered/successfully-registered.component.ts:14]: done

**Pull request merged**